### PR TITLE
Add dequantize-softmax option to TFL-to-TOSA pass

### PR DESCRIPTION
**Pull request by bjacob**
This allows temporarily working around issues with quantized softmax as discussed in https://github.com/google/iree/issues/8974 .

This introduces a new boolean pass option, `dequantize_softmax`. The default `false` preserves current behavior. Setting it to `true` results in dequantizing Softmax ops, which breaks bit-for-bit compatibility but is the right compromise for certain downstreams (IREE). Also note that as Softmax isn't currently a TOSA op, the TOSA spec isn't mandating anything here -- so the bit-for-bit accuracy that we are breaking here is at most breaking only an invariant in the TFLite import pipeline, not in the TOSA spec.

While the performance gap originally reported in https://github.com/google/iree/issues/8974 has shrunk as observed in https://github.com/google/iree/issues/9170, dequantizing softmax is still a 20% end-to-end speedup on a mobilebert benchmark even after that (used to be a > 2x speedup).  In addition to performance, other angles are being discussed in https://github.com/google/iree/issues/8974 around generally how to handle something like a quantized softmax, how early to introduce implementation details in the IR, etc.

### CODE REVIEW

**Comment by bjacob** [comment]
I wasn't able to add @sjarus as reviewer so here's a FYI.

**Comment by sjarus** [comment]
> I wasn't able to add @sjarus as reviewer so here's a FYI.

Of course, I'd be happy to take a look at this. We have never been able to add reviewers properly either 🙂

**Comment by bjacob** [comment]
Talked briefly with @rsuderman - going to refactor this PR so that instead of introducing a boolean option to the TFL-to-TOSA pass, it introduces a new separate TFlite-to-TFLite preprocessing pass. The TFL-to-TOSA pipeline will then be able to run that greedily, first. This will hopefully make this PR less intrusive.

**Comment by sjarus** [comment]
No problem, just ping me when you're done and I'll take a look.

File: tensorflow/compiler/mlir/tosa/transforms/dequantize_tfl_softmax.cc [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+struct TosaDequantizeTFLSoftmaxPattern : public RewritePattern {
+  explicit TosaDequantizeTFLSoftmaxPattern(MLIRContext* context)
+      : RewritePattern(TFL::SoftmaxOp::getOperationName(), 1, context) {}
+  LogicalResult matchAndRewrite(Operation* op,
+                                PatternRewriter& rewriter) const override;
+};
+
+LogicalResult TosaDequantizeTFLSoftmaxPattern::matchAndRewrite(
+    Operation* op, PatternRewriter& rewriter) const {
+  fprintf(stderr, "enter matchAndRewrite\n");
```
Comment (sjarus) [inline tensorflow/compiler/mlir/tosa/transforms/dequantize_tfl_softmax.cc:48]: Leftover debugging output ?

File: tensorflow/compiler/mlir/tosa/transforms/dequantize_tfl_softmax.cc [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+      RankedTensorType::get(input_type.getShape(), rewriter.getF32Type());
+  Value dequantized_input = rewriter.create<TFL::DequantizeOp>(
+      loc, dequantized_input_type, tfl_softmax_op.input());
+  Value dequantized_softmax_output = rewriter.create<TFL::SoftmaxOp>(
+      loc, dequantized_input_type, dequantized_input, tfl_softmax_op.beta());
+  Type qtype = tfl_softmax_op.output().getType();
+  rewriter.replaceOpWithNewOp<TFL::QuantizeOp>(tfl_softmax_op, qtype,
+                                               dequantized_softmax_output,
+                                               mlir::TypeAttr::get(qtype));
+  fprintf(stderr, "dequantized, return success from matchAndRewrite\n");
```
Comment (sjarus) [inline tensorflow/compiler/mlir/tosa/transforms/dequantize_tfl_softmax.cc:68]: And here.

File: tensorflow/compiler/mlir/tosa/transforms/dequantize_tfl_softmax.cc [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+};
+
+LogicalResult TosaDequantizeTFLSoftmaxPattern::matchAndRewrite(
+    Operation* op, PatternRewriter& rewriter) const {
+  fprintf(stderr, "enter matchAndRewrite\n");
+  TFL::SoftmaxOp tfl_softmax_op = cast<TFL::SoftmaxOp>(op);
+  RankedTensorType input_type =
+      tfl_softmax_op.input().getType().cast<RankedTensorType>();
+  if (!input_type.getElementType().isa<mlir::quant::QuantizedType>()) {
+    fprintf(stderr, "not quantized, return failure from matchAndRewrite\n");
```
Comment (sjarus) [inline tensorflow/compiler/mlir/tosa/transforms/dequantize_tfl_softmax.cc:53]: Nit: llvm::errs() stream appears to be preferred to stderr.

**Review by sjarus — COMMENTED** [review]
Not sure I'm supposed to review this yet, but I just scanned the changes as a sanity check and though I'd point out some things we want to avoid carrying over if you're working on further commits.

File: tensorflow/compiler/mlir/tosa/transforms/dequantize_tfl_softmax.cc (hunk 3 above)
Comment (bjacob) [reply tensorflow/compiler/mlir/tosa/transforms/dequantize_tfl_softmax.cc]: sorry, meant to drop these printfs. thanks for catching.

**Comment by bjacob** [comment]
Explanation for why the `#include` of the `passes.h.inc` is moved from the N `.cc` files to the `passes.h` file.

As I added a pass dependent on the `TFLDialect`, I now needed every place including `passes.h.inc` to have a forward-declaration of `TFLDialect`. That was more convenient to do if `passes.h.inc` was included in a single place, in `passes.h`, so I only needed the new forward-declaration in that single file.

**Comment by sjarus** [comment]
Thanks for the changes and the cleanup of passes.h.inc is nice too! LGTM from me. 

The only potential point of awkwardness is a TFL->TFL pass sitting in the TOSA subfolder and potential maintainability. However, that's really a question I pose back to the maintainers here, i.e. is that ok ? What you needed to do in terms of TFL dialect fwd decl reflects that. However, I can see why it helps to have this here because it's part of the TOSA import path support and placing it here may make it easier to leverage externally.

**Comment by bjacob** [comment]
Yes, that's a good summary of the compromise here, as discussed internally. It seemed best to have a separate TFL-to-TFL pattern to make this minimally intrusive into the LegalizeTFL pass, and as this pattern is currently only used as part of TFL-to-TOSA import pipelines, it still made sense to have it here for now.

**Comment by sjarus** [comment]
Yup that's how I'd have concluded on it too, so SGTM!

**Review by rsuderman — APPROVED** [review]
